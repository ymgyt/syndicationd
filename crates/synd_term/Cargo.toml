[package]
authors.workspace     = true
categories.workspace  = true
description.workspace = true
edition.workspace     = true
homepage.workspace    = true
keywords.workspace    = true
license.workspace     = true
readme.workspace      = true
repository.workspace  = true

include = ["src/**/*", "CHANGELOG.md", "categories.toml"]
name    = "synd-term"
version = "0.2.3"

[[bin]]
name = "synd"
path = "src/main.rs"

[dependencies]
synd-auth = { path = "../synd_auth", version = "0.2.1" }
synd-feed = { path = "../synd_feed", version = "0.3.1" }
synd-o11y = { path = "../synd_o11y", version = "0.1.5" }

anyhow               = { workspace = true }
chrono               = { workspace = true, features = ["std", "now"] }
clap                 = { workspace = true, features = ["derive", "string", "color", "suggestions", "wrap_help", "env", "std"] }
crossterm            = { version = "0.27.0", features = ["event-stream"] }
directories          = "5.0.1"
edit                 = "0.1.5"
futures-util         = "0.3.30"
graphql_client       = { workspace = true }
html2text            = { version = "0.12" }
itertools            = { workspace = true }
nom                  = { version = "7.1.3", default-features = false, features = ["std"] }
nucleo               = "0.5.0"
open                 = "5.1.0"
parse_duration       = { workspace = true }
ratatui              = { version = "=0.26.1" }
reqwest              = { workspace = true }
schemars             = { version = "0.8.16", default-features = false, features = ["derive"] }
serde                = { workspace = true, features = ["derive"] }
serde_json           = "1.0.111"
thiserror            = { workspace = true }
tokio                = { workspace = true, features = ["macros", "rt-multi-thread", "sync", "time"] }
toml                 = { version = "0.8.12", default-features = true }
tracing              = { workspace = true }
tracing-appender     = "0.2.3"
tracing-subscriber   = { workspace = true }
tui-big-text         = "0.4.3"
unicode-segmentation = "1.10.1"
url                  = { workspace = true }
# https://github.com/arkbig/throbber-widgets-tui/pull/5
# throbber-widgets-tui = "0.3.0"

[features]
# Integration test
integration = []

[dev-dependencies]
synd-api  = { path = "../synd_api" }
synd-test = { path = "../synd_test" }

axum-server  = { workspace = true }
kvsd         = { workspace = true }
serial_test  = { version = "3.0.0", default_features = false, features = ["async", "file_locks"] }
tempfile     = "3"
tokio-stream = "0.1.15"

[lints]
workspace = true

[package.metadata.dist]
dist    = true
formula = "synd"
