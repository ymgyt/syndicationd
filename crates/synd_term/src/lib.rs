#![allow(clippy::new_without_default)]
#![warn(rustdoc::broken_intra_doc_links)]

pub mod application;
pub mod auth;
pub mod cli;
pub mod client;
pub(crate) mod command;
pub mod config;
pub mod interact;
pub mod job;
pub mod keymap;
pub mod matcher;
pub mod terminal;
pub mod types;
pub mod ui;

#[cfg(feature = "integration")]
pub mod integration;
