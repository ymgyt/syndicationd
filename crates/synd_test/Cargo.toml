[package]
authors.workspace     = true
categories.workspace  = true
description.workspace = true
edition.workspace     = true
license.workspace     = true
readme.workspace      = true
repository.workspace  = true

name    = "synd-test"
publish = false
version = "0.1.0"

[dependencies]
synd-auth = { path = "../synd_auth" }

anyhow       = { workspace = true }
axum         = { workspace = true }
chrono       = { workspace = true }
futures-util = { workspace = true }
headers      = { workspace = true }
jsonwebtoken = { workspace = true }
kvsd         = { workspace = true }
octocrab     = { workspace = true }
serde        = { workspace = true }
serde_json   = { workspace = true }
tempfile     = { workspace = true }
tokio        = { workspace = true, features = ["rt-multi-thread", "net", "time"] }
tracing      = { workspace = true }


[lints]
workspace = true
