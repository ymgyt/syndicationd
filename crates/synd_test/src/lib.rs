use std::path::PathBuf;

pub mod jwt;
pub mod kvsd;
pub mod mock;

pub const TEST_EMAIL: &str = "ymgyt@ymgyt.io";
pub const TEST_USER_ID: &str = "899cf3fa5afc0aa1";

pub fn certificate() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join(".dev")
        .join("self_signed_certs")
        .join("certificate.pem")
}

pub fn certificate_buff() -> String {
    std::fs::read_to_string(certificate()).unwrap()
}

pub fn private_key() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join(".dev")
        .join("self_signed_certs")
        .join("private_key.pem")
}

pub fn private_key_buff() -> Vec<u8> {
    std::fs::read(private_key()).unwrap()
}

pub fn temp_dir() -> tempfile::TempDir {
    tempfile::TempDir::new().unwrap()
}
