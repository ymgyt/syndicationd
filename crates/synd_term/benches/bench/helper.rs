use std::time::Duration;

use ratatui::backend::TestBackend;
use synd_term::{
    application::{Application, Cache, Config},
    client::Client,
    config::Categories,
    terminal::Terminal,
    ui::theme::Theme,
};
use url::Url;

pub fn init_app() -> Application {
    let terminal = {
        let backend = TestBackend::new(120, 40);
        let terminal = ratatui::Terminal::new(backend).unwrap();
        Terminal::with(terminal)
    };

    let client = {
        Client::new(
            Url::parse("http://dummy.ymgyt.io").unwrap(),
            Duration::from_secs(10),
        )
        .unwrap()
    };

    let config = { Config::default().with_idle_timer_interval(Duration::from_micros(1)) };

    let cache = { Cache::new(tempfile::TempDir::new().unwrap().into_path()) };

    Application::builder()
        .terminal(terminal)
        .client(client)
        .categories(Categories::default_toml())
        .config(config)
        .cache(cache)
        .theme(Theme::default())
        .build()
}
