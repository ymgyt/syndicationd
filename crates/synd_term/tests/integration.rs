#[cfg(feature = "integration")]
mod test {
    use std::path::{Path, PathBuf};

    use chrono::{TimeZone, Utc};
    use synd_term::{
        application::{Config, Features},
        auth::Credential,
        key, shift,
    };
    use synd_test::temp_dir;

    mod helper;
    use crate::test::helper::{resize_event, test_config, TestCase};

    #[tokio::test(flavor = "multi_thread")]
    async fn login_with_github() -> anyhow::Result<()> {
        helper::init_tracing();

        let test_case = TestCase {
            mock_port: 6000,
            synd_api_port: 6001,
            kvsd_port: 47379,
            terminal_col_row: (120, 30),
            device_flow_case: "case1",
            cache_dir: temp_dir().into_path(),
            ..Default::default()
        };
        let mut application = test_case.init_app().await?;
        let (tx, mut event_stream) = helper::event_stream();

        {
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "initial login prompt",
            }, {
                insta::assert_debug_snapshot!("initial_login", application.buffer());
            });
        }

        {
            // push enter => start auth flow
            tx.send(key!(enter));
            application.event_loop_until_idle(&mut event_stream).await;
            insta::with_settings!({
                description => "show device flow code",
            },{
                insta::assert_debug_snapshot!("device_flow_prompt", application.buffer());
            });
        }

        {
            // polling device access token complete
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "initial landing entries",
            },{
                insta::assert_debug_snapshot!("landing_entries", application.buffer());
            });
        }

        {
            // Rotate theme
            tx.send_multi([shift!('t'), shift!('t'), shift!('t')]);
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
        }

        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn login_with_google() -> anyhow::Result<()> {
        helper::init_tracing();

        let test_case = TestCase {
            mock_port: 6010,
            synd_api_port: 6011,
            kvsd_port: 47389,
            terminal_col_row: (120, 30),
            device_flow_case: "case1",
            cache_dir: temp_dir().into_path(),
            ..Default::default()
        };
        let mut application = test_case.init_app().await?;
        let (tx, mut event_stream) = helper::event_stream();

        {
            // push enter => start auth flow
            // Select google then select
            tx.send_multi([key!('j'), key!(enter)]);
            application.event_loop_until_idle(&mut event_stream).await;
            insta::with_settings!({
                description => "show google device flow code",
            },{
                insta::assert_debug_snapshot!("google_device_flow_prompt", application.buffer());
            });
        }

        {
            // polling device access token complete
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "initial landing entries after google login",
            },{
                insta::assert_debug_snapshot!("google_landing_entries", application.buffer());
            });
        }
        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn refresh_expired_google_jwt() -> anyhow::Result<()> {
        let (expired_jwt, expired_at) = synd_test::jwt::google_expired_jwt();
        let test_case = TestCase {
            mock_port: 6040,
            synd_api_port: 6041,
            kvsd_port: 6042,
            terminal_col_row: (120, 30),
            device_flow_case: "case1",
            cache_dir: temp_dir().into_path(),
            ..Default::default()
        }
        .with_credential(Credential::Google {
            id_token: expired_jwt,
            refresh_token: "dummy".into(),
            expired_at,
        });

        let mut application = test_case.init_app().await?;
        let (_tx, mut event_stream) = helper::event_stream();

        {
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "after_refreshing_expired_google_jwt",
            },{
                insta::assert_debug_snapshot!("refresh_expired_google_jwt_landing", application.buffer());
            });
        }

        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    #[allow(clippy::too_many_lines)]
    async fn subscribe_then_unsubscribe() -> anyhow::Result<()> {
        helper::init_tracing();

        let test_case = TestCase {
            mock_port: 6020,
            synd_api_port: 6021,
            kvsd_port: 47399,
            terminal_col_row: (120, 30),
            interactor_buffer_fn: Some(|case: &TestCase| {
                vec![
                    format!(
                        "should rust http://localhost:{mock_port}/feed/twir_atom",
                        mock_port = case.mock_port
                    ),
                    // edit requirement from should to must
                    format!(
                        "must rust http://localhost:{mock_port}/feed/twir_atom",
                        mock_port = case.mock_port
                    ),
                    // internal error
                    format!(
                        "may rust http://localhost:{mock_port}/feed/error/internal",
                        mock_port = case.mock_port
                    ),
                    format!(
                        "may rust http://localhost:{mock_port}/feed/error/malformed",
                        mock_port = case.mock_port
                    ),
                ]
            }),
            ..Default::default()
        }
        .already_logined();

        let mut application = test_case.init_app().await?;
        let (tx, mut event_stream) = helper::event_stream();

        {
            // Move tab to feeds
            tx.send(key!(tab));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "after feeds tab move",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_landing_feeds", application.buffer());
            });
        }

        {
            // Subscribe
            tx.send(key!('a'));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "after parsing editor buffer for subscribe",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_after_editor_parse", application.buffer());
            });
        }

        {
            // Open feed. TODO: assert interactor
            tx.send(key!(enter));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
        }

        {
            // Edit feed
            tx.send(key!('e'));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "after edit requirement from should to must",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_after_edit", application.buffer());
            });
        }

        {
            // Prompt unsubscribe popup and move selection
            tx.send_multi([key!('d'), key!('l'), key!('h')]);
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "unsubscribe popup",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_unsubscribe_popup", application.buffer());
            });
        }

        {
            // Select Yes (assuming Yes is selected)
            tx.send(key!(enter));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "after unsubscribe",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_unsubscribed", application.buffer());
            });
        }

        {
            // Handle the case that the server of the feed user tried to subscribe to is returning a internal error.
            tx.send(key!('a'));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "handle internal error of the feed server",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_feed_server_internal_error", application.buffer());
            });
        }

        {
            // Handle the case that the server of the feed user tried to subscribe to is returning a internal error.
            tx.send(key!('a'));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "handle malformed xml error ",
            },{
                insta::assert_debug_snapshot!("subscribe_then_unsubscribe_malformed_xml_error", application.buffer());
            });
        }

        Ok(())
    }
    #[tokio::test(flavor = "multi_thread")]
    #[allow(clippy::too_many_lines)]
    async fn filter_entries() -> anyhow::Result<()> {
        helper::init_tracing();

        let test_case = TestCase {
            // this port is hard coded in fixtures
            mock_port: 6030,
            synd_api_port: 6031,
            kvsd_port: 47409,
            kvsd_root_dir: PathBuf::from(env!("CARGO_MANIFEST_DIR"))
                .join("tests/fixtures/kvsd/20240609"),
            terminal_col_row: (120, 30),
            config: Config {
                // To test pagination
                entries_per_pagination: 1,
                feeds_per_pagination: 1,
                ..test_config()
            },
            ..Default::default()
        }
        .already_logined();

        let mut application = test_case.init_app().await?;
        let (tx, mut event_stream) = helper::event_stream();

        // Initial fetch
        {
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "filter entries after initial fetch",
            },{
                insta::assert_debug_snapshot!("filter_entries_initial_fetch", application.buffer());
            });
            // Cover move
            tx.send_multi([key!('j'), key!('g'), key!('e'), key!('g'), key!('g')]);
        }

        // Move Feed tab
        {
            tx.send(key!(tab));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "feed tab after initial fetch",
            },{
                insta::assert_debug_snapshot!("filter_entries_initial_fetch_feed", application.buffer());
            });
            // Cover move_last and move_first
            tx.send_multi([key!('g'), key!('e'), key!('g'), key!('g')]);
            // Move back
            tx.send(key!(tab));
        }

        {
            // Open entry. TODO: assert interactor
            tx.send(key!(enter));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
        }

        // Filter by requirement
        {
            // Change requirement to MUST
            tx.send_multi([key!('h'), key!('h')]);
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "entris after changing requirement to must",
            },{
                insta::assert_debug_snapshot!("filter_entries_req_must_entries", application.buffer());
            });
            // Change requirement to MAY
            tx.send_multi([key!('l'), key!('l')]);
        }

        // Filter by category
        {
            // Enable category filter and activate first category
            tx.send_multi([key!('c'), key!('-'), key!('a')]);
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "entris after enable category filter",
            },{
                insta::assert_debug_snapshot!("filter_entries_category_filter_entries", application.buffer());
            });
            // Enable all category
            tx.send_multi([key!('+'), key!(esc)]);
        }

        // Filter by keyword
        {
            // Enter keyword 'rust 549'
            tx.send_multi([
                key!('/'),
                key!('r'),
                key!('u'),
                key!('s'),
                key!('t'),
                key!(' '),
                key!('5'),
                key!('4'),
                key!('9'),
            ]);
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "entris after keyword search",
            },{
                insta::assert_debug_snapshot!("filter_entries_keyword_search_entries", application.buffer());
            });
            // Clear keyword
            tx.send_multi([
                key!(backspace),
                key!(backspace),
                key!(backspace),
                key!(backspace),
                key!(backspace),
                key!(backspace),
                key!(backspace),
                key!(backspace),
                key!(esc),
            ]);
        }
        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn resize_terminal_to_zero() -> anyhow::Result<()> {
        let (mut col, mut row) = (120, 30);
        let test_case = TestCase {
            mock_port: 6050,
            synd_api_port: 6051,
            kvsd_port: 6052,
            terminal_col_row: (col, row),
            ..Default::default()
        }
        .already_logined();

        let mut application = test_case.init_app().await?;
        let (tx, mut event_stream) = helper::event_stream();

        loop {
            col /= 2;
            row /= 2;
            if col == 0 && row == 0 {
                break;
            }
            // Assert that app do not panic
            tx.send(resize_event(col, row));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
        }
        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn unauthorized() -> anyhow::Result<()> {
        let test_case = TestCase {
            mock_port: 6060,
            synd_api_port: 6061,
            kvsd_port: 6062,
            terminal_col_row: (120, 30),
            ..Default::default()
        }
        .with_credential(Credential::Github {
            // Use invalid token to mock unauthorized
            access_token: synd_test::GITHUB_INVALID_TOKEN.to_owned(),
        });

        let mut application = test_case.init_app().await?;
        let (_tx, mut event_stream) = helper::event_stream();

        // Assert unauthorized error message
        {
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "unauthorized error message(port is ignorable)",
            },{
                insta::assert_debug_snapshot!("unauthorized_error_message", application.buffer());
            });
        }

        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn github_notifications() -> anyhow::Result<()> {
        helper::init_tracing();

        let test_case = TestCase {
            mock_port: 6070,
            synd_api_port: 6071,
            kvsd_port: 6072,
            terminal_col_row: (120, 30),
            // Enable github notification features
            config: Config {
                features: Features {
                    enable_github_notification: true,
                },
                ..test_config()
            },
            now: Some(Utc::with_ymd_and_hms(&Utc, 2024, 5, 5, 8, 0, 0).unwrap()),
            ..Default::default()
        }
        .already_logined();

        let mut application = test_case.init_app().await?;
        let (tx, mut event_stream) = helper::event_stream();

        {
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;
            insta::with_settings!({
                description => "github notifications initial",
            },{
                insta::assert_debug_snapshot!("gh_notifications_init", application.buffer());
            });
        }

        // TODO
        /*
        {
            // Done
            tx.send(key!('d'));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;

            // Unsubscribe
            tx.send(key!('u'));
            application
                .wait_until_jobs_completed(&mut event_stream)
                .await;

            insta::with_settings!({
                description => "github notifications mark as done",
            },{
                insta::assert_debug_snapshot!("gh_notifications_mark_as_done", application.buffer());
            });
        }
        */

        Ok(())
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn cli_commands() -> anyhow::Result<()> {
        helper::init_tracing();
        let test_case = TestCase {
            // If the mock server on port 7000 is not functioning properly on macOS
            // it might be due to the AirPlay Receiver.
            // try uncheking System Preferences > AirPlay Receiver > On to resolve the issue.
            mock_port: 7000,
            synd_api_port: 7001,
            kvsd_port: 7002,
            cache_dir: temp_dir().into_path(),
            ..Default::default()
        }
        .already_logined();

        test_case.init_app().await?;

        check_command_test(test_case.synd_api_port);
        export_command_test(test_case.synd_api_port, &test_case.cache_dir);
        term_command_test(&test_case.cache_dir, &test_case.log_path);
        // Exec clean last
        clean_command_test(&test_case.cache_dir);

        Ok(())
    }

    fn check_command_test(api_port: u16) {
        let mut cmd = assert_cmd::Command::cargo_bin("synd").unwrap();

        cmd.args([
            "check",
            "--endpoint",
            &format!("https://localhost:{api_port}"),
        ])
        .assert()
        .success();

        cmd.arg("--format=json").assert().success();
    }

    fn export_command_test(api_port: u16, cache_dir: &Path) {
        let mut cmd = assert_cmd::Command::cargo_bin("synd").unwrap();

        cmd.args([
            "export",
            "--endpoint",
            &format!("https://localhost:{api_port}"),
            "--cache-dir",
            &cache_dir.display().to_string(),
        ])
        .assert()
        .success();

        cmd.arg("--print-schema").assert().success();
    }

    fn clean_command_test(cache_dir: &Path) {
        let mut cmd = assert_cmd::Command::cargo_bin("synd").unwrap();

        cmd.args(["clean", "--cache-dir", &cache_dir.display().to_string()])
            .assert()
            .success();
    }

    fn term_command_test(cache_dir: &Path, log_path: &Path) {
        let mut cmd = assert_cmd::Command::cargo_bin("synd").unwrap();

        // Nix do not allow to create log file in user directory
        cmd.args([
            "--dry-run",
            "--cache-dir",
            &cache_dir.display().to_string(),
            "--log",
            &log_path.display().to_string(),
        ])
        .assert()
        .success();
    }
}
