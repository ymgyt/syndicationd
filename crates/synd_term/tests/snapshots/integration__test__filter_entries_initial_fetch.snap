---
source: crates/synd_term/tests/integration.rs
description: filter entries after initial fetch
expression: application.buffer()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 120, height: 30 },
    content: [
        "  Syndicationd                                                                                     󱉯 Entries    󰑫 Feeds ",
        "  󰈶 Filter    MAY  󰭎                                                                                                   ",
        "   Search                                                                                                              ",
        "                                                                                                                        ",
        "  Published   Entry 1/34                                                         Feed                               Req ",
        " 2024-06-03  󰭎 2024-06-03                                                       observability news                 SHD▐",
        "  2024-05-29   This Week in Rust 549                                            This Week in Rust                  MST▐",
        "  2024-05-27  󰭎 2024-05-27                                                       observability news                 SHD▐",
        "  2024-05-22   This Week in Rust 548                                            This Week in Rust                  MST▐",
        "  2024-05-20  󰭎 2024-05-20                                                       observability news                 SHD▐",
        "  2024-05-15   This Week in Rust 547                                            This Week in Rust                  MST ",
        "  2024-05-13  󰭎 2024-05-13                                                       observability news                 SHD ",
        "  2024-05-08   This Week in Rust 546                                            This Week in Rust                  MST ",
        "  2024-05-06  󰭎 2024-05-06                                                       observability news                 SHD ",
        "  2024-04-29  󰭎 2024-04-29                                                       observability news                 SHD ",
        "  2024-04-22  󰭎 2024-04-22                                                       observability news                 SHD ",
        "  2024-04-15  󰭎 2024-04-15                                                       observability news                 SHD ",
        "  2024-04-08  󰭎 2024-04-08                                                       observability news                 SHD ",
        "  2024-04-01  󰭎 2024-04-01                                                       observability news                 SHD ",
        "  2024-03-25  󰭎 2024-03-25                                                       observability news                 SHD ",
        "────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────",
        "  󰯂 Entry     2024-06-03                                                                                                ",
        "  󰏌 URL       https://buttondown.email/o11y.news/archive/2024-06-03/                                                    ",
        "   Published 2024-06-03 20:00 (+00:00)                                                                                 ",
        "                                                                                                                        ",
        "  󱙓 Summary                                                                                                             ",
        "  ## Booking.com’s o11y platform                                                                                        ",
        "                                                                                                                        ",
        "  Santanu Sahoo shared [Events: The 4th pillar of Booking.com’s Observability platform][1] where he describes           ",
        "                              Tab:󰹳  j/k:󰹹  gg:󱞧  ge:󱞥  c:  /:  h/l:  Ent:󰏌  r:󰑓  q:                                ",
    ],
    styles: [
        x: 0, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 14, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 99, y: 0, fg: Rgb(255, 160, 122), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 108, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 10, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 17, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 19, y: 1, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 20, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 23, y: 1, fg: Rgb(247, 76, 0), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 24, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 2, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 10, y: 2, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 0, y: 4, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD | UNDERLINED,
        x: 0, y: 5, fg: Rgb(255, 160, 122), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 119, y: 5, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 0, y: 6, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 6, fg: Rgb(247, 76, 0), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 6, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 7, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 7, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 8, fg: Rgb(247, 76, 0), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 8, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 9, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 9, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 10, fg: Rgb(247, 76, 0), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 10, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 11, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 11, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 12, fg: Rgb(247, 76, 0), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 12, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 13, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 13, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 14, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 14, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 15, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 15, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 16, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 16, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 17, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 17, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 18, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 18, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 19, fg: White, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 19, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 21, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 9, y: 21, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 22, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 7, y: 22, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 23, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 13, y: 23, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 25, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD | UNDERLINED,
        x: 11, y: 25, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 30, y: 29, fg: Rgb(111, 93, 99), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 90, y: 29, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
    ]
}
